common/src/chainstate/utxo.rs

```diff
@@ -0,0 +1,462 @@
+        self.height = value;
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Encode, Decode)]
+pub struct UtxoEntry {
+    /// None when utxo has been spent.
+    utxo: Option<Utxo>,
+    /// The utxo entry is dirty when this version is different from the parent.
+    is_dirty: bool,
+    /// The utxo entry is fresh when the parent does not have this utxo
+    is_fresh: bool,
+}
+
+impl UtxoEntry {
+    pub fn new(utxo: Utxo, is_fresh: bool, is_dirty: bool) -> Self {
+        Self {
+            utxo: Some(utxo),
+            is_dirty,
+            is_fresh,
+        }
+    }
+
+    pub(in crate::chainstate) fn new_spent(is_fresh: bool, is_dirty: bool) -> Self {
+        Self {
+            utxo: None,
+            is_dirty,
+            is_fresh,
+        }
+    }
+
+    fn utxo_mut(&mut self) -> Option<&mut Utxo> {
+        self.utxo.as_mut()
+    }
+
+    pub fn is_dirty(&self) -> bool {
+        self.is_dirty
+    }
+
+    pub fn is_fresh(&self) -> bool {
+        self.is_fresh
+    }
+
+    pub fn is_spent(&self) -> bool {
+        self.utxo.is_none()
+    }
+
+    pub fn utxo(&self) -> Option<Utxo> {
+        self.utxo.clone()
+    }
+}
+
+pub trait UtxosView {
+    /// Retrieves utxo.
+    fn get_utxo(&self, outpoint: &OutPoint) -> Option<Utxo>;
+
+    /// Checks whether outpoint is unspent.
+    fn have_utxo(&self, outpoint: &OutPoint) -> bool;
+
+    /// Retrieves the block hash of the best block in this view
+    fn get_best_block_hash(&self) -> Option<H256>;
+
+    /// Estimated size of the whole view (0 if not implemented)
+    fn estimated_size(&self) -> usize;
+
+    /// Performs bulk modification
+    fn batch_write(
+        &mut self,
+        utxos: HashMap<OutPointKey, UtxoEntry>,
+        block_hash: H256,
+    ) -> Result<(), Error>;
+
+    fn derive_cache(&self) -> UtxosCache;
+}
+
+pub fn flush_to_base<T: UtxosView>(
+    cache: UtxosCache,
+    block_hash: H256,
+    base: &mut T,
+) -> Result<(), Error> {
+    base.batch_write(cache.utxos, block_hash)
+}
+
+#[derive(Clone, Default)]
+pub struct UtxosCache<'a> {
+    parent: Option<&'a dyn UtxosView>,
+    current_block_hash: Option<H256>,
+    pub(in crate::chainstate) utxos: HashMap<OutPointKey, UtxoEntry>,
+    memory_usage: usize,
+}
+
+impl<'a> UtxosCache<'a> {
+    /// returns a copy of the UtxoEntry, given the outpoint.
+    fn get_utxo_entry(&self, outpoint: &OutPoint) -> Option<UtxoEntry> {
+        let key = OutPointKey::from(outpoint);
+
+        if let Some(res) = self.utxos.get(&key) {
+            return Some(res.clone());
+        }
+
+        // if utxo is not found in this view, use parent's `get_utxo`.
+        self.parent.map(|parent| {
+            match parent.get_utxo(outpoint) {
+                None => {
+                    // The parent only has an empty entry for this outpoint; we can consider our
+                    // version as fresh.
+                    UtxoEntry::new_spent(true, false)
```

Comment:
I'm confused... trying to understand what this means. So we look in the parent for a utxo, and if it doesn't exist, we create a new one? At least the function name doesn't reflect this, and there's no explanation for why this is the behavior we need.

If we want to indicate that an output is spent, isn't it better to return an enum that says that it's spent?

I really hate same data meaning different things in rust... there's no need to use special values, like 0, to mean something other than the other numbers. So if we `get_utxo_entry` and it doesn't exist, we just say it doesn't exist. The caller can deal with that the way they find appropriate. Is there something I'm missing in this plan?

Reply:
I updated this, removing the `new_spent` function. Also created an enum: `UtxoStatus`.